### PR TITLE
Add TenantCreated Workflow Activity

### PR DESCRIPTION
**Pull request by JoshLefebvre**
Based on the comments in https://github.com/OrchardCMS/OrchardCore/issues/6432.  I would like create a new OnTenantCreatedEvent workflow Activity, however the TenantCreatedEventHandler would be required first. In order for the workflows that I am trying to create, I would need the encoded url in my context.

I am trying to follow the conventions that I see in the OC project to the best of my abilities. However I am wondering why OC puts so much logic in controller actions? I was always taught to keep controllers thin, and add the bulk of the logic to services. Doing so often avoids duplication. The Tenant Admin controller and API controller currently contain a lot of duplicate logic.

I would like to create services or helpers classes for  CreateSetupToken and GetEncodedUrl if possible.

### CODE REVIEW

**Comment by JoshLefebvre** [comment]
So I ended up implementing the entire TenantCreated Workflow Activity. My branch was very much based on @agriffard PR: https://github.com/OrchardCMS/OrchardCore/pull/5878. Perhaps both these workflows could be integrated in together once the issues are resolved.

This is a feature that I require immediately so I have implemented it on my own solution and seems to be working without issues for my current use case.

**Comment by jtkech** [comment]
I will follow it when i will have time ;)

For infos there are other pending PRs that are more or less related

- We are not sure yet how we want to implement shell / tenant events e.g. to keep in sync shells in a distributed env. In my old PR #2247 we used a message bus, now we try to use distributed states instead. In #5249 and #5815 most of the things will be kept in sync, but not yet e.g. shell states.

    We can already have shared storage for shells but we also need to keep them in sync at runtime, so just to say that it is not yet fully done, and that i think it will be done through `ShellHost`.

- The WF `CreateTenantTask` would also have to trigger this event e.g., as said above, through `ShellHost`, and then we may have to check that a `CreateTenantTask` is not executed inline from this new `TenantCreatedEvent` which would result in an infinitive loop. I did the same kind of checkings in #6054 related to WF content events / tasks. But here also we have not fully decided which scenarios we have to prevent or not.

So maybe a little too early to fully know how all of this need to be implemented, but now that rc2 has been shipped i think we will know it quite soon ;)

**Comment by JoshLefebvre** [comment]
Thank you very much for taking the time to write this detailed explanation @jtkech ! I am hoping to start contributing to OC more frequently so I am trying to get more familiar with other parts of the project. 

For the first point I am not sure if I can provide any feedback as I have yet to run OC in a distributed fashion or have any deep knowledge on the subject. For the second point I have already started looking into this and I will try to run some tests to ensure the CreateTenantTask is functioning properly.

Either way I am happy to hear that this will be part of a future milestone  :)

**Comment by agriffard** [comment]
Conflicts to solve.

**Comment by agriffard** [comment]
Error: `The type or namespace name 'DisableTenantTaskDisplay'`, because of the rename of DisplayDriver I guess.

**Comment by agriffard** [comment]
Conflict with sln to fix.

**Comment by mangowi** [comment]
Following :)

**Comment by jptissot** [comment]
@jtkech Do you have an update on how these events would be implemented? A feature I am working on needs to perform an action when a tenant is created and setup. I would like to have the ability to hook into the newly created tenant and modify it's enabled features and settings based on some values set on the admin page of the default tenant. Thank you :)

**Comment by jtkech** [comment]
@jptissot 

The `DistributedShellHostedService` already use some kind of shell host events but at the host level. Hmm, we also have `ISetupEventHandler` that get called at the end of a tenant setup, just before marking it as running, then you would need to get and use a shell scope on the default tenant to retrieve your settings, here by your created tenant itself.

Then, if you need to disable some of the feature for a given tenant, and by the tenant itself as it executes the setup handler, you have to know that the related migrations have already run. Hmm, changing the set of enabled features in the descriptor at the end of the setup, need to be tried but why not ;)

File: src/OrchardCore.Modules/OrchardCore.Tenants/Workflows/Startup.cs [hunk 1]
```diff
@@ -12,9 +14,11 @@ public class Startup : StartupBase
         public override void ConfigureServices(IServiceCollection services)
         {
             services.AddActivity<DisableTenantTask, DisableTenantTaskDisplayDriver>();
-            services.AddActivity<EnableTenantTask, EnableTenantTaskDisplayDriver>();
-            services.AddActivity<CreateTenantTask, CreateTenantTaskDisplayDriver>();
-            services.AddActivity<SetupTenantTask, SetupTenantTaskDisplayDriver>();
+            services.AddActivity<EnableTenantTask, DisableTenantTaskDisplayDriver>();
```
Comment (jptissot) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Workflows/Startup.cs:17]:
```suggestion
            services.AddActivity<EnableTenantTask, EnableTenantTaskDisplayDriver>();
```

File: src/OrchardCore.Modules/OrchardCore.Tenants/Workflows/Startup.cs [hunk 2]
```diff
@@ -12,9 +14,11 @@ public class Startup : StartupBase
...
+            services.AddActivity<CreateTenantTask, DisableTenantTaskDisplayDriver>();
```
Comment (jptissot) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Workflows/Startup.cs:18]:
```suggestion
            services.AddActivity<CreateTenantTask, CreateTenantTaskDisplayDriver>();
```

File: src/OrchardCore.Modules/OrchardCore.Tenants/Workflows/Startup.cs [hunk 3]
```diff
@@ -12,9 +14,11 @@ public class Startup : StartupBase
...
+            services.AddActivity<SetupTenantTask, DisableTenantTaskDisplayDriver>();
```
Comment (jptissot) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Workflows/Startup.cs:19]:
```suggestion
            services.AddActivity<SetupTenantTask, SetupTenantTaskDisplayDriver>();
```

**Comment by Piedone** [comment]
I checked out the PR, going over old ones, and this would be quite useful. Is this something you'd like to revisit @JoshLefebvre or should we close? Note that since then, we also have an API to create tenants. Raising the event from controllers is not necessarily the best place anyway: lower-level services would allow having such an event even if you create a tenant from code.

**Comment by Piedone** [comment]
Closing due to inactivity.

**Pull request closed without merging**